File: Ruby/0003.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    # Hash to store the last index of each character
    char_index = {}
    left = 0
    max_length = 0
  
    s.each_char.with_index do |char, right|
      if char_index.key?(char) && char_index[char] >= left
        # Move the left pointer to the right of the last occurrence of the current character
        left = char_index[char] + 1
      end
  
      # Update the last index of the current character
      char_index[char] = right
      # Calculate the length of the current window and update max_length if needed
      max_length = [max_length, right - left + 1].max
    end
  
    max_length
  end
  

File: PHP/0003.php
class Solution {

/**
 * @param String $s
 * @return Integer
 */
function lengthOfLongestSubstring($s) {
    $n = strlen($s);
    $charMap = []; // Associative array to store the last index of each character
    $maxLength = 0; // Variable to store the maximum length of substring
    $start = 0; // Start index of the sliding window

    for ($end = 0; $end < $n; $end++) {
        $char = $s[$end];
        // If the character is already in the map and its index is within the current window
        if (isset($charMap[$char]) && $charMap[$char] >= $start) {
            // Move the start of the window to the right of the previous index of the current character
            $start = $charMap[$char] + 1;
        }
        // Update the index of the current character
        $charMap[$char] = $end;
        // Calculate the length of the current window and update maxLength if needed
        $maxLength = max($maxLength, $end - $start + 1);
    }

    return $maxLength;
}
}
